dhall/src/Dhall/Core.hs

```diff
@@ -424,7 +424,7 @@ data Expr s a
     | TextShow
     -- | > List                                     ~  List
     | List
-    -- | > ListLit (Just t ) [x, y, z]              ~  [x, y, z] : List t
+    -- | > ListLit (Just t ) [x, y, z]              ~  [x, y, z] : t
```

Comment:
These docs are a bit misleading: In the parser, we only insert the type into the `ListLit` when we have an _empty_ list literal. I'm wondering why though!

Should there be an invariant that we have a `Just t` iff the list is empty?

Reply:
Yes, there should be such an invariant

The more accurate way to encode things in the types would be to have separate `Expr` constructors for empty lists and non-empty lists, like:

```haskell
| EmptyListLit (Expr s a)
| NonEmptyListLit (NonEmptySeq (Expr s a))
```

... but that would be a more disruptive change

Reply:
Actually, another way you could do this is to remove the type annotation entirely from the `List` constructor.  In other words:

```haskell
ListLit (Seq (Expr s a))
```

... and instead the type-checker can enforce that empty list literals are of the form `Annot (ListList []) a`

Reply:
I think having a separate constructor for empty lists is much clearer. Should I open an issue to discuss the idea before I attempt implementing it?

Reply:
@sjakobi: The main reason I suggest using `Annot` is because it's closer to the standard (since now an empty list without an annotation is a type error rather than a syntax error)

Reply:
> (since now an empty list without an annotation is a type error rather than a syntax error)

Do you suggest deviating from the standard here @Gabriel439?

Currently we still have [`tests/parser/failure/unit/ListLitEmptyAnnotation`](https://github.com/dhall-lang/dhall-lang/blob/ee2fe7d9cbd699fc9f40ca4858abcc0f13105324/tests/parser/failure/unit/ListLitEmptyAnnotation.dhall) and the rule `"[" whsp "]" whsp ":" whsp1 annotation-expression`.

Reply:
Oh, you're right, it is still part of the grammar.  I still think splitting out the empty list constructor would be a bit disruptive, though.